Userland/Libraries/LibCore/LocateDb.cpp

```diff
@@ -0,0 +1,43 @@
+#include <LibCore/LocateDb.h>
+
+namespace Core {
+
+LocateDb::LocateDb(String path, LocateDbMode mode)
+{
+    file_handle = fopen(path.characters(), (mode == LocateDbMode::write ? "w" : "r"));
+    byte_offset = 0;
+}
+
+LocateDb::~LocateDb()
+{
+    fclose(file_handle);
+}
+
+void LocateDb::write_header()
+{
+    // Format: magic string + version.
+    String header = String::format("%s|%d\n", "locatedb", 1);
+    fwrite(header.characters(), sizeof(char), header.length(), file_handle);
+}
+
+void LocateDb::write_directory(DirectoryInfo& dir_info)
+{
+    StringBuilder dir_chunk;
+    dir_chunk.append(String::format("%d|%s\n", (uint8_t)RowType::dir_start, dir_info.path.characters()).characters());
+    for (auto& child : dir_info.children) {
+        dir_chunk.append(String::format("%d|%s\n", (uint8_t)(child->type == FileType::file ? RowType::file : RowType::dir), child->name.characters()).characters());
+    }
+    dir_chunk.append(String::format("%d|%s\n", (uint8_t)RowType::dir_end, dir_info.path.characters()).characters());
+    dir_chunk.to_byte_buffer();
+    fwrite(dir_chunk.to_string().characters(), sizeof(char), dir_chunk.to_string().length(), file_handle);
```

Comment:
This looks like it could potentially allocate a huge buffer, maybe write directly to the output file?
if corruption is possible, you could also first output to a temporary file then `rename()` it (with the typical `EXDEV` dance if needed)

Reply:
The idea behind writing in chunks was to avoid hammering the disk with write requests; however, I might be trying to solve a problem that doesn't exist by doing it this way? Maybe it would be a good idea to at least put the StringBuilder on the heap?

Thx for the comments by the way, this was exactly the type of feedback I was looking for :) (will obviously traverse the other issues you had noticed as well)

Reply:
stdio has a bunch of buffers already.
At this level, I don't think you need to worry about disk requests, let the kernel and libc worry about that :P 

As far as StringBuilder goes, it will start with a tiny buffer on the stack for small operations, and switch to heap allocations whenever needed, so...